Break the dependency cycles

gasche opened the pull request:
There were several rather nasty dependency cycles in Xml-light,
which were partially worked around by mutating hook functions
and using Obj.magic, but still prevented some OCaml tooling from
working properly (ocamlbuild for example chokes on incompatible mli/ml
dependency graphs).

The current commits breaks the dependency cycles by:
- adding a module xml_types.mli for all shared datatypes
- adding a module xml_errors.ml for all shared exceptions
- moving the "checking" functionality of Dtd to
  a separate module Dtd_check

Dtd_check exports its "checked" type as a private type, which
guarantees that the only way to build a value is to use its "check"
function. The maps it uses internally are not mutable anymore (which
incurs a slight conversion cost to and from the mutable representation
used in Dtd) to preserve safety when they are passed to the outer
world -- otherwise an user could mutate the maps post-check.

Care was taken to preserve interface compatibility by re-exporting the
declarations of all new modules in the outer modules -- direct users
of Xml, Dtd and XmlParser should see no difference.

This is only a preliminary commit, I would like to do the following
improvements before a merge:
- better specify the module hierarchy in comments
- do a review to make sure interface compatibility is preserved
  (in particular for exceptions)
- use the longer names Xml_light_{types,errors} to avoid module name
  conflicts
- test with OPAM projects using xml-light that it works at runtime

To test that the new organization works, I do
  ocamlbuild xml-light.cma
(I guess I can remove xml-light.mllib in the finalized version of the
patch; the point is _not_ to change the project's build system.)

gasche commented:
I just rebased the branch to use long names (`Xml_light_...`) for the new modules, to make extra sure we avoid introducing naming clashes in user applications. I also pushed two minor additional commit patches.

I would be interested in feedback from @ncannasse or current xml-light users before going any further on this, though. Is there actually interest in merging this change which simplifies the code structure? I understand that the `xml-light` is strictly in maintenance mode, and that there are other (possibly better) OCaml libraries in that space nowadays, but given that xml-light still has users¹, I think it would be nice to have this cleanup for the long-term wins -- for example, avoiding `Obj.magic` means that people auditing their code and dependencies for unsafety risks will not spend time on this library.

¹: [the opam page](http://opam.ocaml.org/packages/xml-light/xml-light.2.4/) currently lists amqp-client, bulletml, dbforge, gpx, gxl-light, ocamleditor, ofx, tcx as dependencies.

gasche commented:
People may be interested by the graph of module dependencies (produced by ocamldoc) for the new organization:

http://gallium.inria.fr/~scherer/tmp/xml-light-module-dependencies.svg

the graph of type dependencies is much less readable:

http://gallium.inria.fr/~scherer/tmp/xml-light-type-dependencies.svg

olafhering commented:
Xml_light_utils is appearently missing.

gasche commented:
Sorry, I just fixed this. It's easy to make this mistake because for some reason the project's `.gitignore` ignores all `.ml`,`.mli` files.

olafhering commented:
@ncannasse are you still maintaining this repository?

ncannasse commented:
Not much tbh.

Le lun. 4 nov. 2019 à 11:01, Olaf Hering <notifications@github.com> a
écrit :

> @ncannasse <https://github.com/ncannasse> are you still maintaining this
> repository?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ncannasse/xml-light/pull/3?email_source=notifications&email_token=AAHZXQCHXZORP3QRZLG4ZT3QR7XGBA5CNFSM4CGOJHSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEC6W6EY#issuecomment-549285651>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHZXQHJ5R43KECM23SS6ELQR7XGBANCNFSM4CGOJHSA>
> .
>

kit-ty-kate commented:
@ncannasse would you be willing to move it to https://github.com/ocaml-community ?

hhugo commented:
> @ncannasse would you be willing to move it to https://github.com/ocaml-community ?

@ncannasse ^^^

hhugo commented:
@gasche, I can try to push this through as I have write access to the repo. Are you still willing to spend time on this ? rebase and fix the makefile ?

olafhering commented:
At this point the only consumer seems to be `haxe` for me, although a few more consumers are shown in opam. Perhaps this pkg should be moved to `HaxeFoundation`?

Either way, as long as it is moving forward anything will do IMO.

gasche commented:
@hhugo I just rebased this PR and fixed the Makefiles.

The original PR had this big "break the dependency cycles" commit, and then my local branch had two smaller refactoring commits as follow-up. I rebased only the "break the dependency cycle" commit which I think is the more important. If you get very enthusiastic about maintaining xml-light I can send the other two commits later, but preferably in a separate PR.

hhugo commented:
merged in #7

hhugo commented:
things had been rebased by @kit-ty-kate already
